src/cockpit/remote.py

```diff
@@ -15,21 +15,185 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
+import asyncio
+import getpass
+import logging
+import os.path
+import re
+import socket
 
-from typing import Dict
+from typing import Dict, Optional, Tuple
 
-from .router import RoutingError, RoutingRule
+from ._vendor import ferny
+from .router import RoutingRule, Router
+from .peer import Peer, PeerError
+
+logger = logging.getLogger(__name__)
+
+
+class PasswordResponder(ferny.InteractionResponder):
+    PASSPHRASE_RE = re.compile(r"Enter passphrase for key '(.*)': ")
+
+    password: Optional[str]
+    error_message: Optional[str]
+    attempts: int
+
+    def __init__(self, password: Optional[str]):
+        self.password = password
+        self.error_message = None
+        self.attempts = 0
+
+    async def do_hostkey(self, reason: str, host: str, algorithm: str, key: str, fingerprint: str) -> bool:
+        try:
+            logger.debug('checking ~/.ssh/known_hosts')
+            with open(os.path.expanduser('~/.ssh/known_hosts')) as file:
```

Comment:
This gets called, always, on every connect, sometimes multiple times, regardless of if the key is already known or not.

The manpage mentions

> Specifies a command to use to obtain a list of host keys, in addition to those listed in UserKnownHostsFile and GlobalKnownHostsFile.  This command is executed after the files have been read.

but ssh does absolutely nothing to inform us about if it found a match in those files or not, and calls the command unconditionally.

I'm glad you flagged this because I also feel a bit uncomfortable about it.  My "full strength" idea here was to check all of the files (which we can query from ssh with `-G`) and also any KnownHostsCommand that the user might have defined for themselves, but I'm not really thrilled about that either.

I think we might have to discuss the possibility of breaking compatibility with the behaviour the front-end expects from us and switch over to using an interaction that's based more on the "Do you accept this key?" prompt via askpass.  It would also let us avoid being the ones responsible for writing to the file.  More parsing, though... :/

In short, I'm not really sure what to tell you at this point other than "ya, this is gonna need to be better".

Reply:
It was a dumb issue: I accidentally changed the `problem='unknown-hostkey'` to `problem='unknown-host'`, which apparently means something completely different to the frontend :)

Reply:
[current run](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18280-20230418-072056-1e5ebd74-fedora-37-pybridge/log.html#307) didn't finish yet, but this is clearly superior. Thanks so much!